source/geometry_model/box.cc

```diff
@@ -316,16 +317,60 @@ namespace aspect
     bool
     Box<dim>::point_is_in_domain(const Point<dim> &point) const
     {
-      AssertThrow(!this->get_parameters().mesh_deformation_enabled ||
-                  this->simulator_is_past_initialization() == false,
-                  ExcMessage("After displacement of the free surface, this function can no longer be used to determine whether a point lies in the domain or not."));
+      // If mesh deformation occurs over time, and we're past the
+      // first timestep (during which no mesh deformation is applied),
+      // we can no longer use the original extents of the model domain
+      // to check whether the point lies in the domain.
+      // Initial mesh deformation is applied during the first
+      // timestep, but this function might be called before the
+      // simulator is initialized and thus before we can ask whether initial
+      // or time-dependent mesh deformation is applied.
+      // Therefore, when any mesh deformation is enabled,
+      // we loop over all the current grid cells to see if the point
+      // lies within the domain.
+      // Do note that this function can be called before mesh deformation
+      // is applied in the first timestep, and therefore there is no guarantee
+      // that the point will lie in the domain after initial mesh deformation.
+      // For example, boundary traction plugins are initialized before
+      // mesh deformation plugins and the initial_lithostatic_pressure
+      // plugin calls this function while the simulator is being initialized.
+      // However, if the simulator has not been initialized,
+      // no mesh deformation has been applied and we can use the geometry model's
+      // extents to check whether a point lies in the domain.
+      if (this->get_parameters().mesh_deformation_enabled &&
+          this->simulator_is_past_initialization())
+        {
+          std::pair<const typename parallel::distributed::Triangulation<dim>::active_cell_iterator,
+              Point<dim>> it =
+                GridTools::find_active_cell_around_point<> (this->get_mapping(), this->get_triangulation(), point);
```

Comment:
I am not sure this is a great idea. First, it is quite expensive, though perhaps that is not so important because the function is not called very often. But more importantly, this creates difficulties in parallel computations. There, the mesh consists of locally owned, ghost, and artificial cells, and these differ between processes in both where they are and in their geometric shape. So one process may be returning `true` from this function when another one is returning `false`. That's awkward.

I thought about a solution, and all I can come up with is if all processes call this function, and then they figure whether at least *one* process had a locally owned cell where the point was found. This is then independent of the shapes of ghost and artificial cells. Of course, this is *even more expensive*, and furthermore it makes this function a collective function that needs to be called from all processes at the same time, and presumably with the same points.

Reply:
> I thought about a solution, and all I can come up with is if all processes call this function, and then they figure whether at least one process had a locally owned cell where the point was found.

Yes, I think this is the only possible solution. We do something similar in source/postprocess/point_values.cc:75-91. There, we use the function `VectorTools::point_value`, but internally that function uses `find_active_cell_around_point` as well. then the result of the function is evaluated and communicated to all processes.